### PR TITLE
ECR: support removing leading indentation and trailing newline

### LINKED ISSUE #2980: ECR doend't support silent (<%- %>) tags?
I am trying to generate some Javascript output from a Crystal template, so I need precise control over the output. But with the current ECR I get newlines all over the place, which makes ECR unsuitable for my use case.

Am I missing anything? Thoughts?

### PR DESCRIPTION
**Pull request by asterite**
Fixes #2980

This is an alternative implementation to #2985 

@beno I hope you don't mind I decided to try to implement this, but I think the needed changes were really few, and there's no need to introduce more classes nor use regex for this. To do this one just has to look at at most one token ahead, and check if its string value needs to be adjusted. Please check the code, I might have made a mistake or maybe I didn't implement it correctly (I tried a few templates and they seemed to work exactly like ERB).

This now supports all of ERB idioms, as you can see in the specs. Additionally, one can also use `<%-= ... %>`, which removes leading indentation from it, and I guess because of how it's done this also works for `<%-# ... %>`.

### CODE REVIEW

**Comment by beno** [comment]
I find this to be a less functional (-% is now illegal to use?)  much more impenetrable and messy implementation that favours an ugly procedural style over OO principles, but if this is what you prefer I guess it'll have to do.

I also wonder if this implementation correctly respects all line semantics, but I'm not  going to check. Thanks for solving this.

File: src/ecr/processor.cr [hunk 1]
```diff
@@ -42,13 +63,40 @@ module ECR
     end
   end
 
-  private def append_loc(str, filename, token)
+  private def do_supress_leading(token, string)
```
Comment (ysbaddaden) [inline src/ecr/processor.cr]: Maybe `suppress_leading_whitespace`?

**Comment by ysbaddaden** [comment]
I agree. Skipping trailing and leading linefeeds can be considered as mere modifiers and we should avoid new types, as well as regular expressions —they're slow and may quickly increase compile time— to achieve this.

Yet, maybe the lexer could foresee a second char, so we could properly parse `-℅>`?

**Comment by asterite** [comment]
@beno I fixed the `-%` issue.

I don't try to use OOP for everything, every programming style has its uses and pros/cons. In this case I want ECR parsing to be super fast, and creating many objects and using regex is slow.

**Pull request merged**